core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java

```diff
@@ -1563,6 +1563,19 @@ public ServerBuilder withVirtualHost(Consumer<? super VirtualHostBuilder> custom
      * @return {@link VirtualHostBuilder} for building the virtual host
      */
     public VirtualHostBuilder virtualHost(String hostnamePattern) {
+        requireNonNull(hostnamePattern, "hostnamePattern");
+
+        final HostAndPort hostAndPort = HostAndPort.fromString(hostnamePattern);
+        final Optional<VirtualHostBuilder> vhost =
+                virtualHostBuilders.stream()
+                                   .filter(v -> !v.defaultVirtualHost() &&
+                                                v.equalsHostnamePattern(hostAndPort.getHost(),
```

Comment:
The `hostname` is normalized thus this won't still find the builder if the name contains unicode.
https://github.com/line/armeria/blob/3adf832543640b9b9a09a6bd6c65cda1c72963a3/core/src/main/java/com/linecorp/armeria/server/VirtualHost.java#L197

So how about validating and normalizing it before we find it?
Then we can just set the `hostnamePattern` and `port` directly to `VirtualHostBuilder` not to call `HostAndPort.fromString()` and validate it again.
```java

```

Reply:
Good idea. But, I don't fully understand your suggestion. Could you check I understand is correct?

> Then we can just set the hostnamePattern and port directly to VirtualHostBuilder not to call HostAndPort.fromString() and validate it again.

1. validate and normalize in `ServerBuilder.virtualHost(hostnamePattern)` 
2. if exists on same hostname, return it 
3. if not exists, `VirtualHostBuilder.hostnamePattern(hostnamePattern, port)` calls
4. `VirtualHostBuilder.hostnamePatter(hostnamePattern, port)` sets directly hostnamePattern and port without validation and normalization

Reply:
> In 4, I'm worried about the hostnamePattern is able to be set without validation somewhere.

That's a good question. I believe that if we can make it package-private method we don't have to worry about it.
If you worry about it, we can also add the validation logic in the method but I don't believe we really need it. 😉

Reply:
Oh, I didn't think about `package-private`! great job 👍 Thank you for the suggestion!